### PR TITLE
Update bam2bax to properly populate pk values (mid/mean)

### PR DESCRIPTION
**Pull request by pbjd**
Requires access to the trace file for proper conversion, otherwise it falls back to simulated ScanData, which will work mechanically but some values in the output will be wrong.  If trace file provided, copies ScanData to both bas.h5 and pls.h5 outputs.

### CODE REVIEW

File: utils/bam2bax/src/Converter.cpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
+#include "Converter.h"
+
+Converter::Converter(Settings & settings)
```
Comment (pb-cdunn) [inline utils/bam2bax/src/Converter.cpp]: Better, if possible:

    Settings const&

File: utils/bam2bax/src/Converter.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            if (not writer_->WriteOneZmw(smrt, ras) or not writer_->Errors().empty()) { break; }
+        }
+        if (not settings_.ignoreQV) writer_->WriteFakeDataSets();
+        for (auto error: writer_->Errors()) { AddErrorMessage(error); }
+    }
+
+    return errors_.empty();
+}
+
+void Converter::MockScanData(PacBio::BAM::ReadGroupInfo& rg) {
```
Comment (pb-cdunn) [inline utils/bam2bax/src/Converter.cpp]: `Fake`, not `Mock`, I think.

File: utils/bam2bax/src/Converter.cpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        std::cout << "Converting BAM to bax.h5." << std::endl;
+        writer_ = new HDFBaxWriter(outfn, bcvers,
+            scanData_->BaseMap(), qvs, Bam2BaxDefaults::Bax_Regions_RegionTypes);
+    } else if (mode == Settings::PulseMode) {
+        std::cout << "Converting BAM to plx.h5." << std::endl;
+        writer_ = new HDFPulseWriter(outfn, bcvers,
+            scanData_->BaseMap(), qvs, Bam2BaxDefaults::Bax_Regions_RegionTypes);
+    } else {
+        std::cerr << "UNKNOWN mode." << settings_.mode << std::endl;
+        // XXX: Throw initialization exception
```
Comment (pb-cdunn) [inline utils/bam2bax/src/Converter.cpp]: Why not at least `throw std::exception()` for now?

File: utils/bam2bax/src/Converter.cpp [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    }
+}
+
+void Converter::SetInverseGain(HDFFile& traceFile) {
+    H5::Group acqGrp = traceFile.hdfFile.openGroup("/ScanData/AcqParams");
+    H5::Attribute aduAttr = acqGrp.openAttribute("AduGain");
+    float igain;
+    H5::DataType* dt = new H5::DataType(H5::PredType::IEEE_F32LE);
+    aduAttr.read(*dt, &igain);
+    HDFPulseWriter* pw = static_cast<HDFPulseWriter*>(writer_);
```
Comment (pb-cdunn) [inline utils/bam2bax/src/Converter.cpp]: Better:

    dynamic_cast<HDFPulseWriter*>(writer_)

**Comment by ylipacbio** [comment]
It might worth adding a tiny test with input trace.h5 file to see if everything works as expected.

File: utils/bam2bax/src/Converter.cpp (hunk 3 above)
Comment (pbjd) [inline utils/bam2bax/src/Converter.cpp]: Yes, forgot to update that.

File: utils/bam2bax/src/Converter.cpp (hunk 4 above)
Comment (pbjd) [inline utils/bam2bax/src/Converter.cpp]: Less portable (requires RTTI), plus goes against the style-guide.  Make a more compelling argument!

Comment (pb-cdunn) [inline utils/bam2bax/src/Converter.cpp]: Just an idea. I noticed that the base class had a virtual destructor already. The `dynamic_cast<>` would return `0` on a type mis-match, which would be noticed on the next line. `static_cast<>` might not cause an error until much later in the program. At least we have `valgrind` on Linux. But it's no big deal to me.

File: utils/bam2bax/src/Converter.cpp (hunk 2 above)
Comment (pbjd) [inline utils/bam2bax/src/Converter.cpp]: Maybe, but fake seems to general here.  What we're doing here is mimicking a real object so outside consumers can interact with it mechanically and finish their business, which seems more like a mock.

**Comment by pbjd** [comment]
RE: tiny test.  Indeed, I need some more test coverage for this, but not sure if I'll add an h5 file to the repo.  Will consider though for a future pull request.

File: utils/bam2bax/src/Converter.cpp (hunk 2 above)
Comment (pb-cdunn) [inline utils/bam2bax/src/Converter.cpp]: Maybe I misread this. It seems like we're using an actual `ScanData` instance, but with fake data inside it. Mocking replaces the methods.

**Comment by pb-cdunn** [comment]
What's really nice is if you can generate the h5 file at runtime for the test. Then very little needs to be added to the repo. But we already have so many repos way over 1MB that I won't object.

**Comment by pbjd** [comment]
RE: generate h5 file.  I was thinking along the same lines, but that will take a bit of time.

**Comment by ylipacbio** [comment]
A tiny trace.h5 with a few zmws would do, or place trace.h5 file under /pbi, (e.g., bam2bax/tests/cram/*.t for an example).

**Pull request merged**